pkg/trace/api/telemetry.go

```diff
@@ -0,0 +1,169 @@
+// is proxied back to the client, while for all aditional endpoints the
+// response is discarded. There is no de-duplication done between endpoint
+// hosts or api keys.
+//
+// Could be extended in the future to allow supporting more product endpoints
+// by simply parametrizing metric tags, and logger names
+type telemetryMultiTransport struct {
+	Transport http.RoundTripper
+	Endpoints []*config.Endpoint
+	Director  func(*http.Request)
+}
+
+func (m *telemetryMultiTransport) roundTrip(req *http.Request, endpoint *config.Endpoint) (*http.Response, error) {
+	tags := []string{
+		fmt.Sprintf("endpoint:%s", endpoint.Host),
+	}
+	defer func(now time.Time) {
+		metrics.Timing("datadog.trace_agent.telemetry_proxy.roundtrip_ms", time.Since(now), tags, 1)
+	}(time.Now())
+
+	req.Host = endpoint.Host
+	req.URL.Host = endpoint.Host
+	req.URL.Scheme = "https"
+	req.Header.Set("DD-API-KEY", endpoint.APIKey)
+
+	m.Director(req)
+
+	resp, err := m.Transport.RoundTrip(req)
+	if err != nil {
+		metrics.Count("datadog.trace_agent.telemetry_proxy.roundtrip.errors", 1, tags, 1)
+	}
+	return resp, err
+}
+
+// NewReverseProxy creates an http.ReverseProxy which will forward requests via transport
+func NewReverseProxy(transport http.RoundTripper, logger *stdlog.Logger) http.Handler {
+	director := func(req *http.Request) {
+		req.Header.Set("Via", fmt.Sprintf("trace-agent %s", info.Version))
+		if _, ok := req.Header["User-Agent"]; !ok {
+			// explicitly disable User-Agent so it's not set to the default value
+			// that net/http gives it: Go-http-client/1.1
+			// See https://codereview.appspot.com/7532043
+			req.Header.Set("User-Agent", "")
+		}
+	}
+	return &httputil.ReverseProxy{
+		Director:  director,
+		ErrorLog:  logger,
+		Transport: transport,
+	}
+}
+
+// RoundTrip sends request first to MainTarget, then sends a copy of main request to every configurged
+// AdditionalTargets receiptient.
+//
+// All requests will be sent irregardless of any errors
+// If any request fails, the error will be logged. Only main target's
+// error will be propagated via return value
+func (m *telemetryMultiTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if len(m.Endpoints) == 1 {
+		return m.roundTrip(req, m.Endpoints[0])
+	}
+	slurp, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, err
+	}
+	newreq := req.Clone(req.Context())
+	newreq.Body = ioutil.NopCloser(bytes.NewReader(slurp))
+	rresp, rerr := m.roundTrip(newreq, m.Endpoints[0])
+	for _, endpoint := range m.Endpoints[1:] {
+		newreq := req.Clone(req.Context())
+		newreq.Body = ioutil.NopCloser(bytes.NewReader(slurp))
+		if resp, err := m.roundTrip(newreq, endpoint); err == nil {
+			// we discard responses for all subsequent requests
+			io.Copy(ioutil.Discard, resp.Body)
+			resp.Body.Close()
+		} else {
+			log.Error(err)
+		}
+	}
+	return rresp, rerr
+}
+
+func telemetryEndpoints(conf config.AgentConfig) (endpoints []*config.Endpoint) {
```

Comment:
Is this used anywhere?

Reply:
Uh sorry - this shouldn't have been commited.